### DISCUSSION TITLE
Rbenv install 3.1.2 fails on Mac 12.6 (21G115)

### DISCUSSION THREAD

**Discussion by besi** [rbenv/ruby-build#2061] [Build failures] [answered]
### Steps to reproduce the behavior
- I did a fresh mac install 12.6
- installed Xcode 14.0.1 (14A400) also install the dependencies and run `xcode-select --install`
- Installed homebrew
- Install openssl: `brew install openssl@3`
- Instaslled rbenv `brew install rbenv`
- Finally I attempted to install the ruby 3.1.2

```
rbenv install --list
2.6.10
2.7.6
3.0.4
3.1.2
...
```

Then `rbenv install 3.1.2 --verbose` and I'm getting:

```
/var/folders/v5/d9y1qs_16nv9x5bsvjp1dv480000gn/T/ruby-build.20221004111747.39897.q0zBZU ~
Downloading ruby-3.1.2.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.1/ruby-3.1.2.tar.gz
curl: (77) error setting certificate verify locations:  CAfile: /usr/local/etc/openssl/certs/cert.pem CApath: none
error: failed to download ruby-3.1.2.tar.gz

BUILD FAILED (macOS 12.6 using ruby-build 20220930)
```

As mentioned in issue #1998 as well as #2003 I tried including the `OPENSSL_CFLAGS ` like so:

    export OPENSSL_CFLAGS=-Wno-error=implicit-function-declaration
    rbenv install 3.1.2 --verbose

Same result: `... error: failed to download ruby-3.1.2.tar.gz ...`

Currently I'm unable to install 3.1.2.
I did try another version with the same error: `error: failed to download ruby-2.7.6.tar.bz2`.

I did also try to use the following strategies and `export`s but to no avail:

    echo insecure >> ~/.curlrc
    export PATH="/opt/homebrew/opt/openssl@3/bin:$PATH"

Any help is appreciated!

**Comment by hsbt** [comment]
> * Also added the path to my ~/.zshrc: `  echo 'export PATH="/usr/local/opt/openssl@3/bin:$PATH"' >> ~/.zshrc`

This instruction is unnecessarily.

**Reply by besi** [reply]
Thanks for your comment. I did remove it.

**Reply by mislav** [reply]
I would also add that this line is unnecessary too:
```sh
export CPPFLAGS="-I/opt/homebrew/opt/openssl@3/include"
```
And if you are installing Ruby 3.1.x or greater, then the version of openssl to link to should be `openssl@3` and not `openssl@1.x`:
```sh
export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@3)"
```
For Ruby < 3.1, you can install `openssl@1.1` with Homebrew and link to that.

**Comment by eregon** [comment]
It's pretty clear certificates on your system (or curl but that's unlikely) are not installed or broken:
```
curl: (77) error setting certificate verify locations:  CAfile: /usr/local/etc/openssl/certs/cert.pem CApath: none
error: failed to download ruby-3.1.2.tar.gz
```

**Reply by mislav** [reply]
Yes, your curl looks broken. @besi: It looks like you might have installed curl via Homebrew. You can try unlinking Homebrew's curl and let the system one be used:
```sh
brew unlink curl
hash -r
which curl
#=> /usr/bin/curl
```

**Reply by besi** [reply]
```
~ which curl
/usr/local/opt/curl/bin/curl
~ brew unlink curl

Unlinking /usr/local/Cellar/curl/7.85.0... 0 symlinks removed.
~ hash -r
~ which curl

~ rbenv install 3.1.2
To follow progress, use 'tail -f /var/folders/v5/d9y1qs_16nv9x5bsvjp1dv480000gn/T/ruby-build.20221004165713.48716.log' or pass --verbose
Downloading openssl-3.0.5.tar.gz...
-> https://www.openssl.org/source/openssl-3.0.5.tar.gz
error: failed to download openssl-3.0.5.tar.gz

~  brew reinstall curl
curl is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

~ which curl
/usr/local/opt/curl/bin/curl


~ curl --version
curl 7.85.0 (x86_64-apple-darwin21.5.0) libcurl/7.85.0 (SecureTransport) OpenSSL/1.1.1q zlib/1.2.11 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.3 libssh2/1.10.0 nghttp2/1.50.0 librtmp/2.3
Release-Date: 2022-08-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

~ brew unlink curl
Unlinking /usr/local/Cellar/curl/7.85.0... 0 symlinks removed.
```

It sais the following:

> `curl` is keg-only, which means it was not symlinked into `/usr/local`,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.


Please note that I can download the specified file without problems when using curl myself:

```
curl https://www.openssl.org/source/openssl-3.0.5.tar.gz --output ~/Desktop/openssl-3.0.5.tar.gz 

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14.3M  100 14.3M    0     0  25.8M      0 --:--:-- --:--:-- --:--:-- 25.9M
```

```
~ brew uninstall curl
Uninstalling /usr/local/Cellar/curl/7.85.0... (504 files, 4MB)


```

OpenSSL informs me about some caveats:

> ==> Caveats
> A CA file has been bootstrapped using certificates from the system
> keychain. To add additional certificates, place .pem files in
> 
>      /usr/local/etc/openssl@3/certs
> 
> and run
>   
>     /usr/local/opt/openssl@3/bin/c_rehash
> 
> openssl@3 is keg-only, which means it was not symlinked into /usr/local,
> because macOS provides LibreSSL.
> 
> If you need to have openssl@3 first in your PATH, run:
> 
>      echo 'export PATH="/usr/local/opt/openssl@3/bin:$PATH"' >> ~/.zshrc
> 
> For compilers to find openssl@3 you may need to set:
>  
>     export LDFLAGS="-L/usr/local/opt/openssl@3/lib"
>     export CPPFLAGS="-I/usr/local/opt/openssl@3/include"
> 

So I tried exporting all the System Roots from my keychain and place it in `/usr/local/etc/openssl@3/certs` and ran `/usr/local/opt/openssl@3/bin/c_rehash` still no love.

**Comment by besi** [comment, marked as answer]
Alright I got it working now:

    mkdir -p  /usr/local/etc/openssl/certs/
    ln -s /usr/local/etc/openssl@3/cert.pem /usr/local/etc/openssl/certs/cert.pem 
    rbenv install 3.1.2
